Provenance (#126)
OCR-D/spec: provenance.md

# Provenance
Provenance is information about entities, activities, and people involved in producing a piece of data or thing, which can be used to form assessments about its quality, reliability or trustworthiness. (Source: [The PROV Data Model (PROV-DM)](https://www.w3.org/TR/prov-dm/))

## Data Model
[The PROV Data Model (PROV-DM)](https://www.w3.org/TR/prov-dm/) is used to store all provenance metadata.
All provenance have to be stored in files. 
There's provenance at the page level, but there's also provenance at the document level.
All files regarding provenance were stored in a subfolder '**metadata**'.

## Format
The workflow provenance is stored in [PROV-XML](https://www.w3.org/TR/prov-xml/). 

### Types
All Activities, Entities belonging to the OCR-D workflow have the same namespace.
#### Namespace
<dl>
  <dt><strong>Prefix</strong></dt>
  <dd>ocrd</dd>

>   <dt><strong>Namespace</strong></dt>

Author: Namespace is the namespace. Prefix is just a shortcut for the namespace.

>   <dt><strong>Namespace</strong></dt>

Reviewer: No. The whole list describes the namespace. The first point gives the namespace prefix, and the second the **namespace name**. You can say _namespace URI_ as well. But there is no such thing like a _namespace namespace_.

  <dd>http://www.ocr-d.de</dd>
</dl>

Type | Data Type | Description
-------- | -------- | --------
Entity   | ocrd:mets | Filename  of METS file
Entity   | ocrd:mets_referencedFile | ID of the file referenced inside METS. 
Entity   | ocrd:parameter_file   | Content of the parameter file.

> Activity   | ocrd:module | Module that was executed 

Reviewer: terminology: see below

```suggestion
Activity   | ocrd:processor | Processor that was executed 
```

Activity   | ocrd:workflow | Workflow that was executed

> 

Author: These are just values of type (xsd:QName) inside the XML. To make them unique and searchable I decided to give them a unique namespace.

> 

Reviewer: I understand that. But from this list of element names it is not at all apparent what the actual schema (using those elements) looks like. At least not for PROV-XML non-experts. Please add a block quote of the XSD text, at least the most relevant parts. (This is like saying what classes you want to use in an OOP implementation but not showing how.)

## Content 
Only the following information is stored for provenance:
(a) **General data**
1. Workflow engine 
  - Label including version
  - Start date
  - End date

(b) **Module Projects**
2. Module 
  - Label (e.g.: ocrd-kraken-binarize_Version 0.1.0, ocrd/core 1.0.0)
  - Start date
  - End date
3. METS file before executing module
4. METS file after executing module
5. ID of the input file(s)
6. ID of output file(s)

> 7. Content of parameter.json (optional)

Author: Optional because there is not always a corresponding file. If the file exists, it must also be saved.

> 7. Content of parameter.json (optional)

Reviewer: I see. But this is ambiguous. Why not simply _require_ a parameter.json, but allow it to be _empty_ (keeping the defaults)?

### Input/Output
All files listed in METS should be referenced in provenance via their file ID.

> File ID MAY be linked to its location. The location may be replaced due to 

Reviewer: ```suggestion
A file _may_ be linked to its location (URL). The location may be replaced due to 
```

different uses:
1. local files
2. external files

> Files not listed in METS should be linked to their content in provenance. (e.g.: parameter.json)

Reviewer: ```suggestion

All files not referenced in METS must be linked to their content in provenance. (e.g.: parameter.json)
```

### Ingest Workspace to OCR-D Repositorium
At least before ingesting into repository/LTA, the entire provenance must be stored in one file (metadata/ocrd_provenance.xml) to make the provenance searchable.
Therfore all the provenance files are merged into one big file.
This file replaces all provenance files stored in subfolder 'metadata'

## Example
The file structure could look like this after a workflow with 4 steps has been executed.
```
metadata/
   |
   +-- mets.xml.'workflowid'_0000
   |
   +-- mets.xml.'workflowid'_0001
   |
   +-- mets.xml.'workflowid'_0002
   |
   +-- mets.xml.'workflowid'_0003
   |
   +-- mets.xml.'workflowid'_0004
   |
   +-- ocrd_provenance.xml
   |
   +-- provenance_'workflowid'.xml (optional)

> ```

Reviewer: This example is staggeringly incomplete. It would be effective as an example only with the actual _content_ of the named provenance XML files.

Author: It's just a little example to show where the files are stored and how the structure looks like. The schema of the provenance XML files are defined by https://www.w3.org/TR/2013/NOTE-prov-xml-20130430/. It's not helpful to show unreadable XML.

Reviewer: Unreadable XML?

The structure is somewhat informative in itself, granted. But you have different kinds of files here (with different schemas?), and so far no schema was given at all, let alone an instance. If this is meant to be illustrative, you _must_ add an actual example (regardless of how artificial _or_ repetetive it may seem).
